### PR TITLE
semantic: avoid runtime _d_arraysetlength call when .length = 0

### PR DESCRIPTION
**Pull request by ljmf00**
Signed-off-by: Luís Ferreira <contact@lsferreira.net>

---

This simple code
```d
void main()
{
    int[] a;
    a.length = 0;
}
```

Produces this code with a runtime call on semantic analysis:
```d
void main()
{
	int[] a = null;
	_d_arraysetlengthT(a, 0LU);
	return 0;
}
```

The runtime call `_d_arraysetlengthT` does internally slicing if `new_length <= current_length`
```d
if (newlength <= (*p).length)
{
    *p = (*p)[0 .. newlength];
    void* newdata = (*p).ptr;
    return newdata[0 .. newlength];
}
```
For `0` as new length, this expression will always be true, no matter what, so this should be `a = a[0 .. 0]` instead of `_d_arraysetlengthT(a, 0LU)`

I'm not quite sure if this optimization should be made on the semantic analysis but I assumed it because the `ArrayLengthExpr` assignment is converted to `_d_arraysetlengthT` there.

### CODE REVIEW

**Comment by thewilsonator** [comment]
tests?

**Comment by ljmf00** [comment]
> tests?

The behavior of `.length = 0` is already being tested here https://github.com/dlang/dmd/blob/master/test/runnable/test28.d#L130 . I didn't found any test related to `_d_arraysetlengthT` call, so I assumed no furder tests needed.

I'm new to dmd compiler contributions, so, if there's any way to test the behavior of semantic analysis phase, could you guide me, if possible?

**Comment by thewilsonator** [comment]
That test that the behaviour has not changed, not that the implementation is different. You probably want a codegen test here, but those are not very nice to write.

@UplinkCoder , @rainers , anyone else, any better ideas?

**Comment by rainers** [comment]
> @UplinkCoder , @rainers , anyone else, any better ideas?

How about verifying the output of compilation with -vcg-ast?

**Comment by ibuclaw** [comment]
Isn't this what `POST_SCRIPT` is for?  Ensure there is no call to arraysetlength in the disassembled object file (though gdc and ldc is better at this ability in that they compile to assembly).

**Comment by ljmf00** [comment]
> How about verifying the output of compilation with -vcg-ast?

I done a test case for that using `POST_SCRIPT`, can you review?

**Comment by ljmf00** [comment]
Maybe I'm missing something obvious, but the testcase runs locally with success.

**Comment by ljmf00** [comment]
I solved the issue, the test case was malformed because it was testing only 64bit literals.

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
```
 ... fail_compilation/fail_arrayop3c.d -verrors=0 -o-  -fPIC ()
+ check_clean_git
+ git checkout test/compilable/issue17167.sh
+ rm -rf _generated
+ rm -f install.sh
+ rm -f test/compilable/vcg-ast.d.cg
+ make -f posix.mak check-clean-git
ERROR: Found the following residual temporary files.
ERROR: Temporary files should be stored in `test_results` or explicitly removed.
?? test/compilable/vcg-ast-arraylength.d.cg
posix.mak:69: recipe for target 'check-clean-git' failed
make: *** [check-clean-git] Error 1
```

File: test/compilable/extra-files/vcg-ast-arraylength-postscript.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+source tools/common_funcs.sh
+
+# Test if there's call to the runtime for .length = 100
+grep "_d_arraysetlengthT(arr, 100.*)" "${TEST_DIR}/${TEST_NAME}.d.cg"
+# Make sure there's no call to the runtime for .length = 0
+! grep "_d_arraysetlengthT(arr, 0.*)" "${TEST_DIR}/${TEST_NAME}.d.cg"
+# Test if a slice expr is applied for the above case
+grep "arr = arr\[0..0\]" "${TEST_DIR}/${TEST_NAME}.d.cg"
+
+rm_retry "${TEST_DIR}/${TEST_NAME}.d.cg"
```
Comment (thewilsonator) [inline test/compilable/extra-files/vcg-ast-arraylength-postscript.sh:12]: I have no idea why that isn't working.

File: src/dmd/expressionsem.d [hunk 2]
```diff
@@ -9137,32 +9137,44 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+            // use slice expression when arr.length = 0 to avoid runtime call
+            if(exp.e2.isConst() && exp.e2.toUInteger() == 0)
+            {
+                IntervalExp ie = new IntervalExp(ale.loc, exp.e2, exp.e2);
+                Expression se = new SliceExp(ale.loc, ale.e1, ie);
+                Expression as = new AssignExp(ale.loc, ale.e1, se);
+                auto res = as.expressionSemantic(sc);
+                return setResult(res);
+            }
+            else
```
Comment (WalterBright) [inline src/dmd/expressionsem.d:9149]: Do not need `else` when the `if` returns. It makes for a large unnecessary diff.

**Review by WalterBright — DISMISSED** [review]
see my comment

File: src/dmd/expressionsem.d (hunk 2 above)
Comment (ljmf00) [reply src/dmd/expressionsem.d]: Done

File: test/compilable/extra-files/vcg-ast-arraylength-postscript.sh (hunk 1 above)
Comment (ljmf00) [reply test/compilable/extra-files/vcg-ast-arraylength-postscript.sh:12]: It seems an heisenberg issue. I don't know either.

**Comment by thewilsonator** [comment]
You might need to remove the generated file manually , see https://github.com/dlang/dmd/blob/33cde05f531eda125c97fb748520e01a8ff2374f/.circleci/run.sh#L151-L152

**Comment by ljmf00** [comment]
> You might need to remove the generated file manually , see
> 
> https://github.com/dlang/dmd/blob/33cde05f531eda125c97fb748520e01a8ff2374f/.circleci/run.sh#L151-L152

Done. I'll try to fix this later. I don't see a reason for this being hardcoded.

**Comment by thewilsonator** [comment]
could you rebase now that https://github.com/dlang/ci/pull/436 is in. I think that should be it.

**Comment by thewilsonator** [comment]
restarted build kite, added auto merge.

**Comment by MoonlightSentinel** [comment]
This introduced a regression: https://issues.dlang.org/show_bug.cgi?id=21678

---

The problem is that the value of the expression changed from the new length (0)  to the array slice.

Two possible solutions:
- revert this PR
- change the rewrite to yield 0, e.g.

```d
a.length = 0;

(a = a[0 .. 0], 0)
```
